src/unity/python/turicreate/toolkits/image_similarity/image_similarity.py

```diff
@@ -409,3 +409,135 @@ def similarity_graph(self, k=5, radius=None, include_self_edges=False,
+        # by the Core ML model
+        >>> import PIL
+        >>> image = tc.image_analysis.resize(
+                data['image'][0], *reversed(model.input_image_shape))
+        >>> image = PIL.Image.fromarray(image.pixel_data)
+
+        # Calculate distances using the Core ML model
+        >>> ml_model.predict(data={'image': image})
+        {'distance': array([ 0.      , 28.453125, 24.96875 ])}
+        """
+        import numpy as _np
+        import coremltools as _cmt
+        from coremltools.models import datatypes as _datatypes, neural_network as _neural_network
+        from .._mxnet_to_coreml import _mxnet_converter
+        from turicreate.toolkits import _coreml_utils
+
+        # Get the reference data from the model
+        proxy = self.similarity_model.__proxy__
+        reference_data = _np.array(_tc.extensions._nearest_neighbors._nn_get_reference_data(proxy))
+        num_examples, embedding_size = reference_data.shape
+
+        # Get the input and output names
+        input_name = self.feature_extractor.data_layer
+        output_name = 'distance'
+        input_features = [(input_name, _datatypes.Array(*(self.input_image_shape)))]
+        output_features = [(output_name, _datatypes.Array(*(num_examples, )))]
+
+        # Create a neural network
+        builder = _neural_network.NeuralNetworkBuilder(
+            input_features, output_features, mode=None)
+
+        # Convert the feature extraction network
+        mx_feature_extractor = self.feature_extractor._get_mx_module(
+            self.feature_extractor.ptModel.mxmodel,
+            self.feature_extractor.data_layer,
+            self.feature_extractor.feature_layer,
+            self.feature_extractor.context,
+            self.input_image_shape
+        )
+        batch_input_shape = (1, ) + self.input_image_shape
+        _mxnet_converter.convert(mx_feature_extractor, mode=None,
+                                 input_shape={input_name: batch_input_shape},
+                                 builder=builder, verbose=False)
+
+        # To add the nearest neighbors model we add calculation of the euclidean 
+        # distance between the newly extracted query features (denoted by the vector u)
+        # and each extracted reference feature (denoted by the rows of matrix V).
+        # Calculation of sqrt((v_i-u)^2) = sqrt(v_i^2 - 2v_i*u + u^2) ensues.
+        V = reference_data
+        v_squared = (V * V).sum(axis=1)
+
+        feature_layer = self.feature_extractor.feature_layer
+        builder.add_inner_product('v^2-2vu', W=-2 * V, b=v_squared, has_bias=True,
+                                  input_channels=embedding_size, output_channels=num_examples,
+                                  input_name=feature_layer, output_name='v^2-2vu')
+
+        builder.add_elementwise('element_wise-u^2', mode='MULTIPLY',
+                                input_names=[feature_layer, feature_layer],
+                                output_name='element_wise-u^2')
```

Comment:
Can also be done with `add_unary` and `alpha=2` to avoid an unnecessary edge in the graph.